rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,230 @@
+# Distributed DNS Load Balancing
+
+- Feature Name: Distributed DNS Load Balancing
+- Start Date: 2024-01-17
+- RFC PR: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/pull/55)
+- Issue tracking: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/issues/56)
+
+# Summary
+[summary]: #summary
+
+Enable the DNS Operator to manage DNS for one or more hostnames shared across one or more clusters, where a central "hub" cluster is not required or may not be available, and increase the robustness of the DNS Health checks by executing them from all the available clusters.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the DNS record controller has limited functionality for a single cluster, and none for multiple clusters without OCM. Having the DNSRecord controller enabled to manage DNS in these situations significantly eases onboarding users to the advantages of the DNSRecord controller.
+
+# Diagrams
+[diagrams]: #diagrams
+
+- ![Use Cases](0008-distributed-dns-assets%2Fdynamic-dns-use-cases.jpg)
+- ![P2P Health Checks](0008-distributed-dns-assets%2Fdynamic-dns-health-checks.jpg)
+- ![DNS Prunin](0008-distributed-dns-assets%2Fdynamic-dns-dns-pruning.jpg)
+- ![DNS Controller](0008-distributed-dns-assets%2Fdynamic-dns-dns-controller.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster and Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret, any that reference the same zone(s) in the same DNS Provider for the same host will behave as a distributed DNS. In each cluster, when a matching listener host defined in the gateway is found that matches one of the available zones in the provider the addresses from that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways addresses to the same zone file, without destroying each other's values.
+
+### Health Checks
+If health checks are defined in the DNS Policy, these will be created on all clusters. When a health check is failing, a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
```

Comment:
As discussed on a call. Lets remove this from the API for now, use a default of 20% and then allow it to be overridden at the controller level initially. We can look to move to the API if we get enough requests / see enough use cases. Removing it removes configuration and complexity that the user can get wrong in a MC scenerio

Reply:
I guess potentially we could allow it as an annotation on the DNSPolicy and marked as experimental or something also